Make deployment jobs compatible with standalone and integrated mode

amoralej opened the pull request:
We are working in integrating the watcher-operator in the openstack-operator following the same approach that other openstack services operators are doing.

To make the transition smooth, I am proposing to make the jobs compatible with both standalone and integrated mode by implementing following changes:

- Now, the content provider jobs should be rebuilding the watcher-operator and the openstack-operator. For the integration mode, that should make that, when the openstack-operator deployes the watcher-operator it will run using the patched version of it.
- I'm adding logic to find out if the watcher-operator is already deployed when running the deploy_watcher_service.yaml playbook. If the Watcher CRDs are already installed, it means the watcher operator has been installed by the openstack-operator and there is no need to install it standalone.
- I'm also adding logic to find out if the watcher is integrated in the openstackcontrolplane. If so, watcher is deployed by patching the existing openstackcontrolplane. Otherwise, it creates a Watcher CR as it is currently doing.

Note that, at some point we will be able to improve this approach by applying a kustomization to the controlplane in pre_deploy, but at that time is not possible to discover if the watcher is already integrated in the openstackcontrolplane CRD, so we can keep using this post_deployment approach until the integration is finished and propagated to all the branches and environments.

Jira: [OSPRH-16242](https://issues.redhat.com//browse/OSPRH-16242)

amoralej commented:
recheck

raukadah commented:
From [content provider log ](https://logserver.rdoproject.org/6e3/rdoproject.org/6e3ddaa11e144d41a3b869c7b53be9f3/ci-framework-data/logs/ci_script_009_openstack_operator_call.log), I can see watcher-operator image is used with in openstack operator.
```
 skopeo copy docker://38.102.83.155:5001/openstack-k8s-operators/watcher-operator-bundle:827a8a9b1c9822c546c4c3b988e60df681d4aeb7 dir:tmp/bindata/tmp
Getting image source signatures
Copying blob sha256:5931c42002678302050b728c3d14b884cce6acd8c700d3768e19b4b9325263e9
Copying blob sha256:d44a0ef923c0818f8e79c27229a4c0ec914d8fc9035b0a739c65e169b4c89e68
Copying blob sha256:74df6485ad5795fde49ab1a918209ee69680f3f4bcb5ffa4535d2b3055701aae
Copying config sha256:e09de00f9fb8533ae7d935f7c97de47e217026508206e4da5f99e85bae4c10c1
Writing manifest to image destination
+ break
+ extract_bundle tmp/bindata/tmp bindata/
+ local IN_DIR=tmp/bindata/tmp
+ local OUT_DIR=bindata/
++ file tmp/bindata/tmp/5931c42002678302050b728c3d14b884cce6acd8c700d3768e19b4b9325263e9 tmp/bindata/tmp/74df6485ad5795fde49ab1a918209ee69680f3f4bcb5ffa4535d2b3055701aae tmp/bindata/tmp/d44a0ef923c0818f8e79c27229a4c0ec914d8fc9035b0a739c65e169b4c89e68 tmp/bindata/tmp/e09de00f9fb8533ae7d935f7c97de47e217026508206e4da5f99e85bae4c10c1 tmp/bindata/tmp/manifest.json tmp/bindata/tmp/version
++ grep gzip
++ cut -f 1 -d :
+ for X in $(file ${IN_DIR}/* | grep gzip | cut -f 1 -d ':')
+ tar xvf tmp/bindata/tmp/5931c42002678302050b728c3d14b884cce6acd8c700d3768e19b4b9325263e9 -C bindata//
tests/
tests/scorecard/
tests/scorecard/config.yaml
+ for X in $(file ${IN_DIR}/* | grep gzip | cut -f 1 -d ':')
+ tar xvf tmp/bindata/tmp/74df6485ad5795fde49ab1a918209ee69680f3f4bcb5ffa4535d2b3055701aae -C bindata//
manifests/
manifests/watcher-operator-controller-manager-metrics-service_v1_service.yaml
manifests/watcher-operator-metrics-reader_rbac.authorization.k8s.io_v1_clusterrole.yaml
manifests/watcher-operator-watcherapi-editor-role_rbac.authorization.k8s.io_v1_clusterrole.yaml
manifests/watcher-operator-watcherapi-viewer-role_rbac.authorization.k8s.io_v1_clusterrole.yaml
manifests/watcher-operator-webhook-service_v1_service.yaml
manifests/watcher-operator.clusterserviceversion.yaml
manifests/watcher.openstack.org_watcherapis.yaml
manifests/watcher.openstack.org_watcherappliers.yaml
manifests/watcher.openstack.org_watcherdecisionengines.yaml
manifests/watcher.openstack.org_watchers.yaml
+ for X in $(file ${IN_DIR}/* | grep gzip | cut -f 1 -d ':')
+ tar xvf tmp/bindata/tmp/d44a0ef923c0818f8e79c27229a4c0ec914d8fc9035b0a739c65e169b4c89e68 -C bindata//
metadata/
metadata/annotations.yaml
```
and from[ EDPM logs](https://logserver.rdoproject.org/1e8/rdoproject.org/1e8fbf870a9a48589a92d912efa0a02a/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack-operators/all_resources.log), I can see watcher-operator is deployed from openstack-operator 
```
pod/watcher-operator-controller-manager-77f8c566ff-9qdx4              2/2     Running     0          72m
```
and
from watcher deploy log file https://logserver.rdoproject.org/1e8/rdoproject.org/1e8fbf870a9a48589a92d912efa0a02a/controller/ci-framework-data/logs/ci_script_013_run_deploy_watcher.log it looks good from first glance.

amoralej commented:
check-rdo

raukadah commented:
Results without openstack-operator depends-on and with Depends-On: https://github.com/openstack-k8s-operators/ci-framework/pull/3075

[On meta content provider ](https://logserver.rdoproject.org/7d3/rdoproject.org/7d34c35f07844a11bf0cf6817e3d054e/job-output.txt) failing task skipped now as watcher operator is not integrated.
```
TASK [operator_build : watcher-operator - Get latest commit when no PR is provided chdir={{ operator.src }}, _raw_params=git show-ref --head --hash head] ***
2025-06-17 11:18:00.557599 | controller | Tuesday 17 June 2025  11:18:00 -0400 (0:00:00.060)       0:07:51.039 **********
2025-06-17 11:18:00.557628 | controller | changed: [localhost]
2025-06-17 11:18:00.607552 | controller |
2025-06-17 11:18:00.607636 | controller | TASK [operator_build : watcher-operator - Set pr_sha to be used as image tag pr_sha={{ operator.pr_sha | default(git_head_out.stdout | trim) }}] ***
2025-06-17 11:18:00.607654 | controller | Tuesday 17 June 2025  11:18:00 -0400 (0:00:00.229)       0:07:51.268 **********
2025-06-17 11:18:00.607683 | controller | ok: [localhost]
2025-06-17 11:18:00.644036 | controller |
2025-06-17 11:18:00.644132 | controller | TASK [operator_build : watcher-operator - Update the go.mod file using latest commit if no PR is provided chdir={{ cifmw_operator_build_meta_src }}, _raw_params=go mod edit -replace {{ operator_api_path }}={{ operator_api_path }}@{{ pr_sha }}
2025-06-17 11:18:00.644150 | controller | go mod tidy
2025-06-17 11:18:00.644164 | controller | if [ -d ./apis ]; then
2025-06-17 11:18:00.644178 | controller |   pushd ./apis/
2025-06-17 11:18:00.644192 | controller |   go mod edit -replace {{ operator_api_path }}={{ operator_api_path }}@{{ pr_sha }}
2025-06-17 11:18:00.644245 | controller |   go mod tidy
2025-06-17 11:18:00.644259 | controller |   popd
2025-06-17 11:18:00.644272 | controller | fi
2025-06-17 11:18:00.644285 | controller | ] ***
2025-06-17 11:18:00.644299 | controller | Tuesday 17 June 2025  11:18:00 -0400 (0:00:00.050)       0:07:51.318 **********
2025-06-17 11:18:00.644327 | controller | skipping: [localhost]
2025-06-17 11:18:11.472775 | controller |
```
On EDPM job side, [we have seperate index for openstack operator and watcher operator ](https://logserver.rdoproject.org/553/rdoproject.org/553654a5fbe74522ba1cef36897e0192/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack-operators/all_resources.log)
```
pod/openstack-operator-controller-manager-8489cf7468-fthzq            2/2     Running     0          38m
pod/openstack-operator-controller-operator-64d74fbd7b-9tmpt           2/2     Running     0          38m
pod/openstack-operator-index-76ndr                                    1/1     Running     0          39m
pod/ovn-operator-controller-manager-9f78645d5-vtmp4                   2/2     Running     0          38m
pod/placement-operator-controller-manager-58f798889d-vthpm            2/2     Running     0          38m
pod/rabbitmq-cluster-operator-manager-67ff8584d-hqqql                 1/1     Running     0          38m
pod/swift-operator-controller-manager-7779c57cf7-xgxcd                2/2     Running     0          38m
pod/telemetry-operator-controller-manager-76656fdd44-84vng            2/2     Running     0          38m
pod/test-operator-controller-manager-5f58c9d867-v7fqs                 2/2     Running     0          38m
pod/watcher-operator-controller-manager-75598496f8-glhfb              2/2     Running     0          4m21s
pod/watcher-operator-index-7pfbc                                      1/1     Running     0          4m49s
```
It means that with above ci-framework pr, standalone watcher validation job works fine.

amoralej commented:
> Results without openstack-operator depends-on and with Depends-On: [openstack-k8s-operators/ci-framework#3075](https://github.com/openstack-k8s-operators/ci-framework/pull/3075)
> 
> [On meta content provider ](https://logserver.rdoproject.org/7d3/rdoproject.org/7d34c35f07844a11bf0cf6817e3d054e/job-output.txt) failing task skipped now as watcher operator is not integrated.
> 
> ```
> TASK [operator_build : watcher-operator - Get latest commit when no PR is provided chdir={{ operator.src }}, _raw_params=git show-ref --head --hash head] ***
> 2025-06-17 11:18:00.557599 | controller | Tuesday 17 June 2025  11:18:00 -0400 (0:00:00.060)       0:07:51.039 **********
> 2025-06-17 11:18:00.557628 | controller | changed: [localhost]
> 2025-06-17 11:18:00.607552 | controller |
> 2025-06-17 11:18:00.607636 | controller | TASK [operator_build : watcher-operator - Set pr_sha to be used as image tag pr_sha={{ operator.pr_sha | default(git_head_out.stdout | trim) }}] ***
> 2025-06-17 11:18:00.607654 | controller | Tuesday 17 June 2025  11:18:00 -0400 (0:00:00.229)       0:07:51.268 **********
> 2025-06-17 11:18:00.607683 | controller | ok: [localhost]
> 2025-06-17 11:18:00.644036 | controller |
> 2025-06-17 11:18:00.644132 | controller | TASK [operator_build : watcher-operator - Update the go.mod file using latest commit if no PR is provided chdir={{ cifmw_operator_build_meta_src }}, _raw_params=go mod edit -replace {{ operator_api_path }}={{ operator_api_path }}@{{ pr_sha }}
> 2025-06-17 11:18:00.644150 | controller | go mod tidy
> 2025-06-17 11:18:00.644164 | controller | if [ -d ./apis ]; then
> 2025-06-17 11:18:00.644178 | controller |   pushd ./apis/
> 2025-06-17 11:18:00.644192 | controller |   go mod edit -replace {{ operator_api_path }}={{ operator_api_path }}@{{ pr_sha }}
> 2025-06-17 11:18:00.644245 | controller |   go mod tidy
> 2025-06-17 11:18:00.644259 | controller |   popd
> 2025-06-17 11:18:00.644272 | controller | fi
> 2025-06-17 11:18:00.644285 | controller | ] ***
> 2025-06-17 11:18:00.644299 | controller | Tuesday 17 June 2025  11:18:00 -0400 (0:00:00.050)       0:07:51.318 **********
> 2025-06-17 11:18:00.644327 | controller | skipping: [localhost]
> 2025-06-17 11:18:11.472775 | controller |
> ```
> 
> On EDPM job side, [we have seperate index for openstack operator and watcher operator ](https://logserver.rdoproject.org/553/rdoproject.org/553654a5fbe74522ba1cef36897e0192/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack-operators/all_resources.log)
> 
> ```
> pod/openstack-operator-controller-manager-8489cf7468-fthzq            2/2     Running     0          38m
> pod/openstack-operator-controller-operator-64d74fbd7b-9tmpt           2/2     Running     0          38m
> pod/openstack-operator-index-76ndr                                    1/1     Running     0          39m
> pod/ovn-operator-controller-manager-9f78645d5-vtmp4                   2/2     Running     0          38m
> pod/placement-operator-controller-manager-58f798889d-vthpm            2/2     Running     0          38m
> pod/rabbitmq-cluster-operator-manager-67ff8584d-hqqql                 1/1     Running     0          38m
> pod/swift-operator-controller-manager-7779c57cf7-xgxcd                2/2     Running     0          38m
> pod/telemetry-operator-controller-manager-76656fdd44-84vng            2/2     Running     0          38m
> pod/test-operator-controller-manager-5f58c9d867-v7fqs                 2/2     Running     0          38m
> pod/watcher-operator-controller-manager-75598496f8-glhfb              2/2     Running     0          4m21s
> pod/watcher-operator-index-7pfbc                                      1/1     Running     0          4m49s
> ```
> 
> It means that with above ci-framework pr, standalone watcher validation job works fine.

Right:

https://logserver.rdoproject.org/553/rdoproject.org/553654a5fbe74522ba1cef36897e0192/controller/ci-framework-data/logs/ci_script_013_run_deploy_watcher.log

```
TASK [Install Watcher Operator in standalone mode output_dir={{ cifmw_basedir }}/artifacts, chdir={{ watcher_repo }}, script=make watcher, extra_args={'CATALOG_IMAGE': "{{ watcher_catalog_image | default('quay.io/openstack-k8s-operators/watcher-operator-index:latest') }}", 'WATCHER_API_CI_IMAGE': '{{ _registry_url }}/openstack-watcher-api:{{ _tag }}', 'WATCHER_DECISION_ENGINE_CI_IMAGE': '{{ _registry_url }}/openstack-watcher-decision-engine:{{ _tag }}', 'WATCHER_APPLIER_CI_IMAGE': '{{ _registry_url }}/openstack-watcher-applier:{{ _tag }}'}] ***
Tuesday 17 June 2025  16:20:11 +0000 (0:00:00.027)       0:00:00.595 ********** 
Follow script's output here: /home/zuul/ci-framework-data/logs/ci_script_014_install_watcher_operator_in.log
changed: [localhost]

TASK [Check if a Watcher CR is already created _raw_params=oc get watcher -A -o name | grep -q watcher] ***
Tuesday 17 June 2025  16:21:01 +0000 (0:00:50.922)       0:00:51.518 ********** 
changed: [localhost]

TASK [Add watcher_is_deployed fact, true if a Watcher CR exist, otherwise false watcher_is_deployed={{ watcher_deployed.rc == 0 }}] ***
Tuesday 17 June 2025  16:21:02 +0000 (0:00:00.412)       0:00:51.931 ********** 
ok: [localhost]

TASK [Check if Watcher is fully integrated in the OpenStackControlPlane _raw_params=oc explain openstackcontrolplane.spec|grep -q -w "watcher"] ***
Tuesday 17 June 2025  16:21:02 +0000 (0:00:00.032)       0:00:51.963 ********** 
changed: [localhost]

TASK [Add watcher_in_ctlplane fact, true if Watcher is in the controlplane CRD watcher_in_ctlplane={{ watcher_controlplane_spec.rc == 0 }}] ***
Tuesday 17 June 2025  16:21:02 +0000 (0:00:00.381)       0:00:52.345 ********** 
ok: [localhost]

TASK [Set watcher_cr_file to the one with container images watcher_cr_file_override={{ watcher_cr_file_full }}] ***
Tuesday 17 June 2025  16:21:02 +0000 (0:00:00.038)       0:00:52.383 ********** 
skipping: [localhost]

TASK [Deploy Watcher service using standalone Watcher CR output_dir={{ cifmw_basedir }}/artifacts, chdir={{ watcher_repo }}, script=make watcher_deploy, extra_args={'WATCHER_SAMPLE_CR_PATH': "{{ watcher_cr_file_override | default( watcher_cr_file | default('ci/watcher_v1beta1_watcher.yaml')) }}", 'WATCHER_API_CI_IMAGE': '{{ _registry_url }}/openstack-watcher-api:{{ _tag }}', 'WATCHER_DECISION_ENGINE_CI_IMAGE': '{{ _registry_url }}/openstack-watcher-decision-engine:{{ _tag }}', 'WATCHER_APPLIER_CI_IMAGE': '{{ _registry_url }}/openstack-watcher-applier:{{ _tag }}'}] ***
Tuesday 17 June 2025  16:21:02 +0000 (0:00:00.034)       0:00:52.417 ********** 
Follow script's output here: /home/zuul/ci-framework-data/logs/ci_script_015_deploy_watcher_service_using.log
changed: [localhost]

TASK [Deploy Watcher service by patching an existing OpenStackControlplane _raw_params=set -ex
CTLPLANE=$(oc get openstackcontrolplane -n openstack -o jsonpath="{range .items[*]}{@.metadata.name}{end}")
oc patch openstackcontrolplane $CTLPLANE -n openstack --type merge --patch-file $CTLPLANE_PATCH
oc wait openstackcontrolplane $CTLPLANE -n openstack --for condition=Ready --timeout=600s
oc wait watcher watcher -n openstack --for condition=Ready --timeout=600s
] ***
Tuesday 17 June 2025  16:21:52 +0000 (0:00:49.406)       0:01:41.824 ********** 
skipping: [localhost]
```

raukadah commented:
recheck

raukadah commented:
recheck

raukadah commented on ci/scenarios/kuttl.yml:
```diff
@@ -22,3 +22,6 @@ post_install_operators_kuttl_from_operator:
         combine({ 'watcher_repo': watcher_repo })
       }}
+
+# Enable observability operator
+cifmw_deploy_obs: true
```
cifmw_deploy_obs var is used in ci-framework to deploy obs: https://github.com/openstack-k8s-operators/ci-framework/blob/ad6505a6417e9ab40ad91542ba86eb2eaae26e46/playbooks/02-infra.yml#L85 . It is integrated with deploy_edpm.yaml playbook used in upstream EDPM job and also in architecture repo.

Sounds good to use this var.
And, I think
we need to drop https://github.com/openstack-k8s-operators/watcher-operator/blob/main/ci/playbooks/deploy_cluster_observability_operator.yaml and https://github.com/openstack-k8s-operators/watcher-operator/blob/main/ci/scenarios/edpm.yml#L7 in favor of above.

raukadah replied:
As discussed on slack, change related to cifmw_deploy_obs will come in seperate pr. It is not blocking the existing one. It is good to merge!

viroel commented:
/lgtm

looks good, we will touch this again soon, once we start to support only the integrated mode, the standalone bits can be removed. Thanks Alfredo.

raukadah commented:
/cherry-pick 18.0-fr3
